### DISCUSSION TITLE
how to update the ForeignKey instance?

### DISCUSSION THREAD

**Discussion by ccurvey** [codingjoe/joeflow#77] [Q&A]
I'm probably overcomplicating this.  

I have a domain model that looks like this:

``` python 
class Purchase(models.Model):
     customer_name = models.CharField(max_length=100)
     invoice_number = models.CharField(max_length=100, null=True, blank=True)
```

And I have a workflow that looks like this:

```python
class PurchaseWorkflow(Workflow):
    purchase = models.ForeignKey(Purchase, on_delete=models.CASCADE...)
```

I would like to create a human task that will allow someone to type the invoice number into a form, and then have the invoice number saved on the `Purchase`. 

The only way I could make this work was to add "invoice_number" to the Workflow, then copy the value over to the Purchase by extending the `form_valid` method, like so:

```python

class InvoiceView(tasks.UpdateView):
    def form_valid(self, *args, **kwargs):
        purchase = self.object.purchase
        purchase.invoice_number = self.object.invoice_number
        purchase.save()
        return super().form_valid(*args, **kwargs)

class PurchaseWorkflow(Workflow):
    purchase = models.ForeignKey(Purchase, on_delete=models.CASCADE...)
    invoice_number = models.CharField(max_length=100, null=True, blank=True)

    add_invoice = InvoiceView(
        fields=['invoice_number'],
        success_url=reverse_lazy("pizza:list-invoices"),
    )
```

But this feels hacky.  Is this the pattern I should be using, or is there a better way?

I was trying to create my own ModelForm with some extra fields, but I can't seem to get a reference to the form from within `form_valid()`

**Comment by codingjoe** [comment]
Hi @ccurvey,

Interesting scenario. Is there a reason why you separated the purchase from the workflow? If they are tightly intertwined, maybe it's better to only have one model. Or is there a reason to duplicate data, that I am missing? In general, I would recommend keeping workflows pretty self-contained (like event-driven programming). Then again, this framework has room for various solutions.

In your case, you could also have a follow-up machine task, that updates the value on a related object. It might be more verbose, about the behavior or your workflow and your architecture.

Please let me know if you have any further questions or need more suggestions.

Cheers!
Joe

**Reply by ccurvey** [reply]
> Is there a reason why you separated the purchase from the workflow? 

I'd like to keep my domain objects (in this case, Purchase) separate from the workflow, as the same domain object may need to be used in several workflows, or even used outside a workflow.  A good example would be an Employee.  An employee might go through a "New Hire" workflow, then through the "Performance Review" workflow several times, and finally through a "Departure" workflow. 

My mental pattern is that a workflow would operate on a domain object directly (1), without having to duplicate the field into the workflow.

Hmm, maybe I'm doing this in the wrong place?  Maybe I should be setting the `form_class` on the `InvoiceView` to be a custom ModelForm that updates the Purchase in the `save()` method?  (I'm just not quite clear on whether or not the UpdateView will somehow call the `save()` method.

(1) I was so happy to see that you can "start" a workflow with a machine task, passing it a domain object.

**Comment by ccurvey** [comment]
a-ha!  found it!

```python

# A vanilla ModelForm related to the Purchase (the domain object)
class InvoiceForm(forms.ModelForm):
    class Meta:
        model = Purchase
        fields = ['invoice_number']

# Subclass tasks.UpdateView so that we can extend form_valid
class InvoiceView(tasks.UpdateView):
    form_class = InvoiceForm
    model = Purchase

    def form_valid(self, *args, **kwargs):
        # the InvoiceView works with the Workflow instance in "object", so we can
        # get our Purchase object from that
        purchase = self.object.purchase

        # there is no "form" attribute/property on the view, so we use "get_form()"
        form = self.get_form()

        # even though the form has been validated, there is no "cleaned_data" attribute.
        # we can either use the "data" attribute, or re-validate
        purchase.invoice_number = form.data['invoice_number']

        # save our related object, and let the workflow processing continue
        purchase.save()

        return super().form_valid(*args, **kwargs)
```
